pkg/controllers/migration/controller.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package migration
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/awslabs/operatorpkg/object"
+	"github.com/awslabs/operatorpkg/singleton"
+	"github.com/samber/lo"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+)
+
+type Controller[T client.Object] struct {
+	kubeClient client.Client
+}
+
+func NewController[T client.Object](client client.Client) *Controller[T] {
+	return &Controller[T]{
+		kubeClient: client,
+	}
+}
+
+func (c *Controller[T]) Reconcile(ctx context.Context) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "migration")
+
+	// create object list
+	o := object.New[T]()
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(object.GVK(o))
+	if err := c.kubeClient.List(ctx, list); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing objects, %w", err)
+	}
+
+	// update annotations on all CRs
+	for _, item := range list.Items {
+		if lo.HasKey(item.GetAnnotations(), v1.StoredVersionMigrated) {
```

Comment:
don't we want to assert that storedversion migrated is true? Could really go either way on this.

Reply:
My thought is presence/absence of the annotation is sufficient to indicate migration.

Reply:
Sounds good to me